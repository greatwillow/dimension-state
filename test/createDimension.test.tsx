import React from 'react'
import renderer from 'react-test-renderer'
import { 
    createDimension, 
    createStore, 
    Provider
} from '../src/index'
import { 
    PlanetManagerReducers,
    PlanetManagerState, 
    planetNamesEnum 
} from './types'

const DIMENSION_STORE_KEY = 'PLANET_MANAGER'
const INITIAL_STATE_PLANET = planetNamesEnum.VENUS


describe("createDimension simple case", () => {
    const store = createStore<PlanetManagerState>({})

    const getInitialState = () => ({
        planet: INITIAL_STATE_PLANET
    })

    const getReducers = () => (state: PlanetManagerState): PlanetManagerReducers => ({
        setPlanet: (planet) => ({
            ...state,
            planet
        })
    })

    const dimension = createDimension({
        dimensionStoreKey: DIMENSION_STORE_KEY,
        initialStateClosure: getInitialState,
        reducersClosure: getReducers,
        asyncActionsClosure: () => ({}),
        selectorsClosure: () => ({}),
        externalDependencies: {},
        store
    })

    const PlanetApp = () => {
        const { planet } = dimension.use()
    
        return (<div>Current planet is: {planet}</div>)
    }

    it('should have proper initial state', () => {
        const { planet } = dimension.store.getState()
        expect(planet).toEqual(INITIAL_STATE_PLANET)
    })

    it("should properly update state", () => {
        const firstUpdatePlanet = planetNamesEnum.MARS
        dimension.reducers.setPlanet(firstUpdatePlanet)
        const { planet } = dimension.store.getState()
        expect(planet).toEqual(firstUpdatePlanet)
    })

    it("should add properly render state", () => {
        const StatefulApp = renderer.create(<Provider store={store}><PlanetApp /></Provider>)

        expect(StatefulApp.toJSON()).toMatchSnapshot()
        renderer.act(() => { dimension.reducers.setPlanet(planetNamesEnum.SATURN) })

        expect(StatefulApp.toJSON()).toMatchSnapshot()

        StatefulApp.unmount()
    })
})