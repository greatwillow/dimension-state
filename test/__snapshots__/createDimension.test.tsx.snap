// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDimension simple case should add properly render state 1`] = `
<div>
  Current planet is: 
  MARS
</div>
`;

exports[`createDimension simple case should add properly render state 2`] = `
<div>
  Current planet is: 
  SATURN
</div>
`;
